File: business-svc/src/main/java/org/zanata/magpie/util/SegmentString.java
package org.zanata.magpie.util;

import org.apache.commons.lang3.StringUtils;
import org.jetbrains.annotations.NotNull;
import org.zanata.magpie.api.dto.LocaleCode;

import java.text.BreakIterator;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import java.util.Optional;

/**
 * @author Alex Eng<a href="mailto:aeng@redhat.com">aeng@redhat.com</a>
 */
public final class SegmentString {

    /**
     * Segment a paragraph into sentences
     */
    public static List<String> segmentString(@NotNull String text,
            Optional<LocaleCode> localeCode) {
        if (StringUtils.isBlank(text)) {
            return new ArrayList<>();
        }

        Locale locale =
                localeCode.isPresent() ? new Locale(localeCode.get().getId()) :
                        Locale.getDefault();
        BreakIterator boundary = BreakIterator.getSentenceInstance(locale);
        boundary.setText(text);
        return extractList(boundary, text);
    }

    private static List<String> extractList(BreakIterator boundary,
            String source) {
        List<String> strings = new ArrayList<>();
        int start = boundary.first();
        for (int end = boundary.next();
                end != BreakIterator.DONE;
                start = end, end = boundary.next()) {
            strings.add(source.substring(start, end));
        }
        return strings;
    }
}

File: business-svc/src/main/java/org/zanata/magpie/util/SegmentString.kt
package org.zanata.magpie.util

import org.zanata.magpie.api.dto.LocaleCode
import java.text.BreakIterator
import java.util.*

/**
 * Segments a paragraph into sentences.
 */
fun segmentBySentences(paragraph: String,
        localeCode: Optional<LocaleCode>): List<String> {
    if (paragraph.isBlank()) return emptyList()

    val locale = if (localeCode.isPresent)
        Locale(localeCode.get().id)
    else
        Locale.getDefault()
    val sentenceIterator = BreakIterator.getSentenceInstance(locale)
    return sentenceIterator.elements(paragraph)
}

// NB this was developed for sentence BreakIterator. YMMV with others.
private fun BreakIterator.elements(text: String): List<String> {
    this.setText(text)
    val elements = ArrayList<String>()
    var startBound = this.first()
    var endBound = this.next()
    while (endBound != BreakIterator.DONE) {
        elements.add(text.substring(startBound, endBound))
        startBound = endBound
        endBound = this.next()
    }
    return elements
}
